src/v/kafka/server/handlers/produce.cc

```diff
@@ -279,37 +279,63 @@ static partition_produce_stages produce_topic_partition(
                         .partition_index = ntp.tp.partition,
                         .error_code = error_code::not_leader_for_partition});
                 }
-                auto stages = partition_append(
-                  ntp.tp.partition,
-                  ss::make_lw_shared<replicated_partition>(
-                    std::move(partition)),
-                  bid,
-                  std::move(reader),
-                  acks,
-                  num_records);
-                return stages.dispatched
-                  .then_wrapped([source_shard, dispatch = std::move(dispatch)](
-                                  ss::future<> f) mutable {
-                      if (f.failed()) {
-                          (void)ss::smp::submit_to(
-                            source_shard,
-                            [dispatch = std::move(dispatch),
-                             e = f.get_exception()]() mutable {
-                                dispatch->set_exception(e);
-                                dispatch.reset();
+                auto f = ss::now();
+                if (partition->is_compacted() && batch.compressed()) {
+                    batch.key_offsets(std::vector<model::key_offset>());
+                    f = storage::internal::decompress_batch(batch).then(
+                      [&batch](model::record_batch&& b) {
+                          return model::for_each_record(
+                            b, [&batch](const model::record& r) {
```

Comment:
I think capturing `&batch` may lead to accessing it after it was freed.

Reply:
It is captured by ref inside f, but after `f` is executed it is captured with move `f.then([.. batch = std::move(batch) ..])` and I thought the latter kept batch alive while `f` is executing. Do think this logic is flawed?

Reply:
I think after the batch is moved reference is invalidated